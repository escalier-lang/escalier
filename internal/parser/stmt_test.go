package parser

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
)

func TestParseStmtNoErrors(t *testing.T) {
	tests := map[string]struct {
		input string
	}{
		// "VarDecl": {
		// 	input: "var x = 5",
		// },
		// "ValDecl": {
		// 	input: "val x = 5",
		// },
		// "ExportValDecl": {
		// 	input: "export val x = 5",
		// },
		// "DeclareValDecl": {
		// 	input: "declare val x",
		// },
		// "ExportDeclareValDecl": {
		// 	input: "export declare val x",
		// },
		// "FunctionDecl": {
		// 	input: "fn foo(a, b) { a + b }",
		// },
		// "FunctionDeclWithReturn": {
		// 	input: "fn foo(a, b) { return a + b }",
		// },
		// "FunctionDeclWithMultipleStmts": {
		// 	input: `fn foo() {
		// 		val a = 5
		// 		val b = 10
		// 		return a + b
		// 	}`,
		// },
		// "ExportFunctionDecl": {
		// 	input: "export fn foo(a, b) { a + b }",
		// },
		// "DeclareFunctionDecl": {
		// 	input: "declare fn foo(a, b)",
		// },
		// "ExportDeclareFunctionDecl": {
		// 	input: "export declare fn foo(a, b)",
		// },
		// "UnicodeVarDecl": {
		// 	input: "val längd = 5",
		// },
		// "UnicodeFunctionDecl": {
		// 	input: "fn до́бра(a, b) { a + b }",
		// },
		// "TypeDecl": {
		// 	input: "type MyType = { x: number, y: string }",
		// },
		// "TypeDeclWithComments": {
		// 	input: `type MyType = Foo
		// 		// Comment
		// 		| Bar
		// 		// Comment
		// 		| Baz`,
		// },
		"TypeDeclWithLeadingPipe": {
			input: `type MyType =
				| Foo
				| Bar
				| Baz`,
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			t.Parallel()
			source := Source{
				Path:     "input.esc",
				Contents: test.input,
			}

			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
			defer cancel()
			parser := NewParser(ctx, source)
			stmt, errors := parser.stmt()

			snaps.MatchSnapshot(t, stmt)
			if len(errors) > 0 {
				fmt.Printf("Error[0]: %#v", errors[0])
			}
			assert.Len(t, errors, 0)
		})
	}
}

func TestParseStmtErrorHandling(t *testing.T) {
	tests := map[string]struct {
		input string
	}{
		"VarDeclMissingIdent": {
			input: "var = 5",
		},
		"VarDeclMissingEquals": {
			input: "var x 5",
		},
		"FunctionDeclMissingIdent": {
			input: `fn () {return 5}`,
		},
		"FunctionDeclMissingBoyd": {
			input: "fn foo(a, b)",
		},
		"FunctionDeclWithIncompleteStmts": {
			input: `fn foo() {
				val a = 
				val b = 5
				return a +
			}`,
		},
		"ParamsMissingOpeningParen": {
			input: "fn foo a, b) { a + b }",
		},
		"ParamsMissingClosingParen": {
			input: "fn foo (a, b { a + b }",
		},
	}

	for name, test := range tests {
		t.Run(name, func(t *testing.T) {
			t.Parallel()
			source := Source{
				Path:     "input.esc",
				Contents: test.input,
			}

			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
			defer cancel()
			parser := NewParser(ctx, source)
			stmt, errors := parser.stmt()

			snaps.MatchSnapshot(t, stmt)
			assert.Greater(t, len(errors), 0)
			snaps.MatchSnapshot(t, errors)
		})
	}
}
