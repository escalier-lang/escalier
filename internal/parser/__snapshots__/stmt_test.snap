
[TestParseStmtErrorHandling/VarDeclMissingIdent - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.VarDecl{
            Kind:    1,
            Pattern: &ast.IdentPat{
                Name:    "",
                Default: nil,
                span:    ast.Span{
                    Start: ast.Location{Line:1, Column:1},
                    End:   ast.Location{Line:1, Column:1},
                },
                inferredType: nil,
            },
            TypeAnn: nil,
            Init:    {
                &ast.LiteralExpr{
                    Lit: &ast.NumLit{
                        Value: 5,
                        span:  ast.Span{
                            Start: ast.Location{Line:1, Column:7},
                            End:   ast.Location{Line:1, Column:8},
                        },
                    },
                    span: ast.Span{
                        Start: ast.Location{Line:1, Column:7},
                        End:   ast.Location{Line:1, Column:8},
                    },
                    inferredType: nil,
                },
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:1, Column:8},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:8},
        },
    },
}
---

[TestParseStmtErrorHandling/FunctionDeclMissingIdent - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.FuncDecl{
            Name: &ast.Ident{
                Name: "",
                span: ast.Span{
                    Start: ast.Location{Line:1, Column:4},
                    End:   ast.Location{Line:1, Column:4},
                },
            },
            FuncSig: ast.FuncSig{
                TypeParams: {
                },
                Params: {
                },
                Return: nil,
                Throws: nil,
            },
            Body: {
                {
                    Stmts: {
                        &ast.ReturnStmt{
                            Expr: {
                                &ast.LiteralExpr{
                                    Lit: &ast.NumLit{
                                        Value: 5,
                                        span:  ast.Span{
                                            Start: ast.Location{Line:1, Column:15},
                                            End:   ast.Location{Line:1, Column:16},
                                        },
                                    },
                                    span: ast.Span{
                                        Start: ast.Location{Line:1, Column:15},
                                        End:   ast.Location{Line:1, Column:16},
                                    },
                                    inferredType: nil,
                                },
                            },
                            span: ast.Span{
                                Start: ast.Location{Line:1, Column:8},
                                End:   ast.Location{Line:1, Column:16},
                            },
                        },
                    },
                    Span: ast.Span{
                        Start: ast.Location{Line:1, Column:7},
                        End:   ast.Location{Line:1, Column:17},
                    },
                },
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:1, Column:17},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:17},
        },
    },
}
---

[TestParseStmtErrorHandling/VarDeclMissingIdent - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:5},
            End:   ast.Location{Line:1, Column:6},
        },
        Message: "Expected a pattern",
    },
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:4},
        },
        Message: "Expected pattern",
    },
}
---

[TestParseStmtErrorHandling/FunctionDeclMissingIdent - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:4},
            End:   ast.Location{Line:1, Column:5},
        },
        Message: "Expected identifier",
    },
}
---

[TestParseStmtErrorHandling/VarDeclMissingEquals - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt](nil)
---

[TestParseStmtErrorHandling/VarDeclMissingEquals - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:7},
            End:   ast.Location{Line:1, Column:8},
        },
        Message: "Expected equals sign",
    },
}
---

[TestParseStmtErrorHandling/FunctionDeclMissingBoyd - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.FuncDecl{
            Name: &ast.Ident{
                Name: "foo",
                span: ast.Span{
                    Start: ast.Location{Line:1, Column:4},
                    End:   ast.Location{Line:1, Column:7},
                },
            },
            FuncSig: ast.FuncSig{
                TypeParams: {
                },
                Params: {
                    &ast.Param{
                        Pattern: &ast.IdentPat{
                            Name:    "a",
                            Default: nil,
                            span:    ast.Span{
                                Start: ast.Location{Line:1, Column:8},
                                End:   ast.Location{Line:1, Column:9},
                            },
                            inferredType: nil,
                        },
                        Optional: false,
                        TypeAnn:  nil,
                    },
                    &ast.Param{
                        Pattern: &ast.IdentPat{
                            Name:    "b",
                            Default: nil,
                            span:    ast.Span{
                                Start: ast.Location{Line:1, Column:11},
                                End:   ast.Location{Line:1, Column:12},
                            },
                            inferredType: nil,
                        },
                        Optional: false,
                        TypeAnn:  nil,
                    },
                },
                Return: nil,
                Throws: nil,
            },
            Body: {
                {
                    Stmts: {
                    },
                    Span: ast.Span{
                        Start: ast.Location{Line:1, Column:13},
                        End:   ast.Location{Line:1, Column:13},
                    },
                },
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:1, Column:13},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:13},
        },
    },
}
---

[TestParseStmtErrorHandling/FunctionDeclMissingBoyd - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:13},
            End:   ast.Location{Line:1, Column:13},
        },
        Message: "Expected an opening brace",
    },
}
---

[TestParseStmtErrorHandling/FunctionDeclWithIncompleteStmts - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.FuncDecl{
            Name: &ast.Ident{
                Name: "foo",
                span: ast.Span{
                    Start: ast.Location{Line:1, Column:4},
                    End:   ast.Location{Line:1, Column:7},
                },
            },
            FuncSig: ast.FuncSig{
                TypeParams: {
                },
                Params: {
                },
                Return: nil,
                Throws: nil,
            },
            Body: {
                {
                    Stmts: {
                        &ast.DeclStmt{
                            Decl: &ast.VarDecl{
                                Kind:    0,
                                Pattern: &ast.IdentPat{
                                    Name:    "a",
                                    Default: nil,
                                    span:    ast.Span{
                                        Start: ast.Location{Line:2, Column:9},
                                        End:   ast.Location{Line:2, Column:10},
                                    },
                                    inferredType: nil,
                                },
                                TypeAnn: nil,
                                Init:    {
                                    &ast.EmptyExpr{
                                        span: ast.Span{
                                            Start: ast.Location{Line:3, Column:5},
                                            End:   ast.Location{Line:3, Column:8},
                                        },
                                        inferredType: nil,
                                    },
                                },
                                export:  false,
                                declare: false,
                                span:    ast.Span{
                                    Start: ast.Location{Line:2, Column:5},
                                    End:   ast.Location{Line:3, Column:8},
                                },
                            },
                            span: ast.Span{
                                Start: ast.Location{Line:2, Column:5},
                                End:   ast.Location{Line:3, Column:8},
                            },
                        },
                        &ast.DeclStmt{
                            Decl: &ast.VarDecl{
                                Kind:    0,
                                Pattern: &ast.IdentPat{
                                    Name:    "b",
                                    Default: nil,
                                    span:    ast.Span{
                                        Start: ast.Location{Line:3, Column:9},
                                        End:   ast.Location{Line:3, Column:10},
                                    },
                                    inferredType: nil,
                                },
                                TypeAnn: nil,
                                Init:    {
                                    &ast.LiteralExpr{
                                        Lit: &ast.NumLit{
                                            Value: 5,
                                            span:  ast.Span{
                                                Start: ast.Location{Line:3, Column:13},
                                                End:   ast.Location{Line:3, Column:14},
                                            },
                                        },
                                        span: ast.Span{
                                            Start: ast.Location{Line:3, Column:13},
                                            End:   ast.Location{Line:3, Column:14},
                                        },
                                        inferredType: nil,
                                    },
                                },
                                export:  false,
                                declare: false,
                                span:    ast.Span{
                                    Start: ast.Location{Line:3, Column:5},
                                    End:   ast.Location{Line:3, Column:14},
                                },
                            },
                            span: ast.Span{
                                Start: ast.Location{Line:3, Column:5},
                                End:   ast.Location{Line:3, Column:14},
                            },
                        },
                        &ast.ReturnStmt{
                            Expr: {
                                &ast.BinaryExpr{
                                    Left: &ast.IdentExpr{
                                        Name:   "a",
                                        Source: nil,
                                        span:   ast.Span{
                                            Start: ast.Location{Line:4, Column:12},
                                            End:   ast.Location{Line:4, Column:13},
                                        },
                                        inferredType: nil,
                                    },
                                    Op:    "+",
                                    Right: &ast.EmptyExpr{
                                        span: ast.Span{
                                            Start: ast.Location{Line:5, Column:4},
                                            End:   ast.Location{Line:5, Column:5},
                                        },
                                        inferredType: nil,
                                    },
                                    span: ast.Span{
                                        Start: ast.Location{Line:4, Column:12},
                                        End:   ast.Location{Line:5, Column:5},
                                    },
                                    inferredType: nil,
                                },
                            },
                            span: ast.Span{
                                Start: ast.Location{Line:4, Column:5},
                                End:   ast.Location{Line:5, Column:5},
                            },
                        },
                    },
                    Span: ast.Span{
                        Start: ast.Location{Line:1, Column:10},
                        End:   ast.Location{Line:5, Column:5},
                    },
                },
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:5, Column:5},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:5, Column:5},
        },
    },
}
---

[TestParseStmtErrorHandling/FunctionDeclWithIncompleteStmts - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:3, Column:5},
            End:   ast.Location{Line:3, Column:8},
        },
        Message: "Expected an expression",
    },
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:5, Column:4},
            End:   ast.Location{Line:5, Column:5},
        },
        Message: "Expected an expression",
    },
}
---

[TestParseStmtErrorHandling/ParamsMissingOpeningParen - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.FuncDecl{
            Name: &ast.Ident{
                Name: "foo",
                span: ast.Span{
                    Start: ast.Location{Line:1, Column:4},
                    End:   ast.Location{Line:1, Column:7},
                },
            },
            FuncSig: ast.FuncSig{
                TypeParams: {
                },
                Params: {
                    &ast.Param{
                        Pattern: &ast.IdentPat{
                            Name:    "a",
                            Default: nil,
                            span:    ast.Span{
                                Start: ast.Location{Line:1, Column:8},
                                End:   ast.Location{Line:1, Column:9},
                            },
                            inferredType: nil,
                        },
                        Optional: false,
                        TypeAnn:  nil,
                    },
                    &ast.Param{
                        Pattern: &ast.IdentPat{
                            Name:    "b",
                            Default: nil,
                            span:    ast.Span{
                                Start: ast.Location{Line:1, Column:11},
                                End:   ast.Location{Line:1, Column:12},
                            },
                            inferredType: nil,
                        },
                        Optional: false,
                        TypeAnn:  nil,
                    },
                },
                Return: nil,
                Throws: nil,
            },
            Body: {
                {
                    Stmts: {
                        &ast.ExprStmt{
                            Expr: &ast.BinaryExpr{
                                Left: &ast.IdentExpr{
                                    Name:   "a",
                                    Source: nil,
                                    span:   ast.Span{
                                        Start: ast.Location{Line:1, Column:16},
                                        End:   ast.Location{Line:1, Column:17},
                                    },
                                    inferredType: nil,
                                },
                                Op:    "+",
                                Right: &ast.IdentExpr{
                                    Name:   "b",
                                    Source: nil,
                                    span:   ast.Span{
                                        Start: ast.Location{Line:1, Column:20},
                                        End:   ast.Location{Line:1, Column:21},
                                    },
                                    inferredType: nil,
                                },
                                span: ast.Span{
                                    Start: ast.Location{Line:1, Column:16},
                                    End:   ast.Location{Line:1, Column:21},
                                },
                                inferredType: nil,
                            },
                            span: ast.Span{
                                Start: ast.Location{Line:1, Column:16},
                                End:   ast.Location{Line:1, Column:21},
                            },
                        },
                    },
                    Span: ast.Span{
                        Start: ast.Location{Line:1, Column:14},
                        End:   ast.Location{Line:1, Column:23},
                    },
                },
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:1, Column:23},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:23},
        },
    },
}
---

[TestParseStmtErrorHandling/ParamsMissingOpeningParen - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:8},
            End:   ast.Location{Line:1, Column:9},
        },
        Message: "Expected an opening paren",
    },
}
---

[TestParseStmtErrorHandling/ParamsMissingClosingParen - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.FuncDecl{
            Name: &ast.Ident{
                Name: "foo",
                span: ast.Span{
                    Start: ast.Location{Line:1, Column:4},
                    End:   ast.Location{Line:1, Column:7},
                },
            },
            FuncSig: ast.FuncSig{
                TypeParams: {
                },
                Params: {
                    &ast.Param{
                        Pattern: &ast.IdentPat{
                            Name:    "a",
                            Default: nil,
                            span:    ast.Span{
                                Start: ast.Location{Line:1, Column:9},
                                End:   ast.Location{Line:1, Column:10},
                            },
                            inferredType: nil,
                        },
                        Optional: false,
                        TypeAnn:  nil,
                    },
                    &ast.Param{
                        Pattern: &ast.IdentPat{
                            Name:    "b",
                            Default: nil,
                            span:    ast.Span{
                                Start: ast.Location{Line:1, Column:12},
                                End:   ast.Location{Line:1, Column:13},
                            },
                            inferredType: nil,
                        },
                        Optional: false,
                        TypeAnn:  nil,
                    },
                },
                Return: nil,
                Throws: nil,
            },
            Body: {
                {
                    Stmts: {
                        &ast.ExprStmt{
                            Expr: &ast.BinaryExpr{
                                Left: &ast.IdentExpr{
                                    Name:   "a",
                                    Source: nil,
                                    span:   ast.Span{
                                        Start: ast.Location{Line:1, Column:16},
                                        End:   ast.Location{Line:1, Column:17},
                                    },
                                    inferredType: nil,
                                },
                                Op:    "+",
                                Right: &ast.IdentExpr{
                                    Name:   "b",
                                    Source: nil,
                                    span:   ast.Span{
                                        Start: ast.Location{Line:1, Column:20},
                                        End:   ast.Location{Line:1, Column:21},
                                    },
                                    inferredType: nil,
                                },
                                span: ast.Span{
                                    Start: ast.Location{Line:1, Column:16},
                                    End:   ast.Location{Line:1, Column:21},
                                },
                                inferredType: nil,
                            },
                            span: ast.Span{
                                Start: ast.Location{Line:1, Column:16},
                                End:   ast.Location{Line:1, Column:21},
                            },
                        },
                    },
                    Span: ast.Span{
                        Start: ast.Location{Line:1, Column:14},
                        End:   ast.Location{Line:1, Column:23},
                    },
                },
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:1, Column:23},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:23},
        },
    },
}
---

[TestParseStmtErrorHandling/ParamsMissingClosingParen - 2]
[]*parser.Error{
    &parser.Error{
        Span: ast.Span{
            Start: ast.Location{Line:1, Column:14},
            End:   ast.Location{Line:1, Column:15},
        },
        Message: "Expected a closing paren",
    },
}
---

[TestParseStmtNoErrors/TypeDeclWithLeadingPipe - 1]
optional.Option[github.com/escalier-lang/escalier/internal/ast.Stmt]{
    &ast.DeclStmt{
        Decl: &ast.TypeDecl{
            Name: &ast.Ident{
                Name: "MyType",
                span: ast.Span{
                    Start: ast.Location{Line:1, Column:6},
                    End:   ast.Location{Line:1, Column:12},
                },
            },
            TypeParams: {
            },
            TypeAnn: &ast.UnionTypeAnn{
                Types: {
                    &ast.TypeRefTypeAnn{
                        Name:     "Foo",
                        TypeArgs: {
                        },
                        span: ast.Span{
                            Start: ast.Location{Line:2, Column:7},
                            End:   ast.Location{Line:2, Column:10},
                        },
                        inferredType: nil,
                    },
                    &ast.TypeRefTypeAnn{
                        Name:     "Bar",
                        TypeArgs: {
                        },
                        span: ast.Span{
                            Start: ast.Location{Line:3, Column:7},
                            End:   ast.Location{Line:3, Column:10},
                        },
                        inferredType: nil,
                    },
                    &ast.TypeRefTypeAnn{
                        Name:     "Baz",
                        TypeArgs: {
                        },
                        span: ast.Span{
                            Start: ast.Location{Line:4, Column:7},
                            End:   ast.Location{Line:4, Column:10},
                        },
                        inferredType: nil,
                    },
                },
                span: ast.Span{
                    Start: ast.Location{Line:2, Column:7},
                    End:   ast.Location{Line:4, Column:10},
                },
                inferredType: nil,
            },
            export:  false,
            declare: false,
            span:    ast.Span{
                Start: ast.Location{Line:1, Column:1},
                End:   ast.Location{Line:1, Column:12},
            },
        },
        span: ast.Span{
            Start: ast.Location{Line:1, Column:1},
            End:   ast.Location{Line:1, Column:12},
        },
    },
}
---
